### DISCUSSION TITLE
Convert internal plugins to use schema customization API

### DISCUSSION THREAD

**Discussion by vladar** [gatsbyjs/gatsby#20069] [Umbrella Discussions]
A follow up for #12272. Now that [createSchemaCustomization API](https://www.gatsbyjs.org/docs/schema-customization/) is stable it is time to start using it for internal Gatsby plugins. Not all plugins are applicable for this API. For such plugins, we should outline why and maybe consider alternatives.

Source plugins:
- [ ] gatsby-source-contentful (#12816)
- [ ] gatsby-source-wordpress (#19292)
- [ ] gatsby-source-drupal
- [ ] gatsby-source-faker
- [ ] gatsby-source-filesystem
- [x] gatsby-source-hacker-news (#20070)
- [x] gatsby-source-lever (#20521)
- [x] gatsby-source-medium (#22377)
- [ ] gatsby-source-mongodb
- [ ] gatsby-source-npm-package-search
- [ ] gatsby-source-wikipedia
- [ ] gatsby-source-shopify

Transformer plugins:
- [x] gatsby-transformer-sharp (#18871)
- [ ] gatsby-transformer-react-docgen (#19514)
- [ ] gatsby-transformer-csv
- [ ] gatsby-transformer-json
- [ ] gatsby-transformer-yaml

## Help wanted
If you are familiar with the [schema customization API](https://www.gatsbyjs.org/docs/schema-customization/) or would like to try it - I encourage you to pick a plugin from the list above which is not being worked on yet, try converting it and publish a PR. I will be happy to help.

If the plugin is not in the list - mention me and name the plugin you would like to work on.

Also if you know community plugins which already use schema customization - please mention them in this issue as well.

**Comment by Powell-v2** [comment]
Hi @vladar, I'd like to help you out with this task! Which plugins are up for grabs?

**Comment by vladar** [comment]
This is great! Thanks for your initiative.

You could pick any plugin from the list above that has no link to existing PR/Epic associated with it. I suggest you mention the one you chose here so that I could update this list to avoid collisions.

One issue though is that some source plugins are too dynamic/complicated (like drupal for instance) so it can be hard to start with those. I think `gatsby-source-medium` could be a low hanging fruit.

You can use [gatsby-plugin-schema-snapshot](https://www.gatsbyjs.org/packages/gatsby-plugin-schema-snapshot) to print a snapshot of a schema, audit and use related type definitions in the plugin.

**Comment by Powell-v2** [comment]
Thanks! I was actually wondering which one of the plugins has the least amount of pitfalls associated with it as I'm just starting out as a contributor. I appreciate your help - will proceed with your suggestion (`gatsby-source-medium`).

**Comment by Qovaros985** [comment]
Hi @vladar, I just created a PR for `gatsby-source-lever`. Could you please guide me, whether that is what we want?

**Comment by vladar** [comment]
@Powell-v2 See also #14765 (just noticed this issue)

**Comment by graygilmore** [comment]
Hey @vladar! I'm just starting to dig into Gatsby and GraphQL but I'd love to lend a hand here if I could. In particular I'm playing around with the Shopify so I might try that one! Let me know if I'm biting off more than I can chew 😄

**Comment by vladar** [comment]
@graygilmore Sure, go for it! I don't think Shopify types are that complex.

One approach that worked for me in the past was to use [`gatsby-plugin-schema-snapshot`](https://www.gatsbyjs.org/packages/gatsby-plugin-schema-snapshot/) on a site with source plugin and use those snapshots to extract types for schema customization API.

Thanks for offering help and feel free to ping me if you need advice!

**Comment by rugggger** [comment]
Hi @vladar , I would like to help for the gatsby-source-drupal plugin.

**Comment by reobin** [comment]
Quick question: Would this allow one to customize the name of a node coming from `gatsby-source-mongodb`, for example?

If for example, nodes are created with the name `mongodbDbnameCollectionname`, but I would like to rename it to something more simple and reusable through various databases (with different names) in the same mongodb cluster.

**Comment by talohana** [comment]
Hi @vladar I would like to help and learn, is gatsby-source-wikipedia beginner friendly?

Thanks!
